feat: add logfmt parse support to the v2 query engine

trevorwhitney opened the pull request:
**What this PR does / why we need it**:

This PR add the logfmt parse stage to the v2 query engine.

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

rfratto commented on pkg/engine/planner/logical/key_collector.go:
```diff
@@ -0,0 +1,141 @@
+// This function is intentionally limited in scope to support optimization of
+// logfmt metric queries in the v2 engine. Support for log queries and other
+// parsers may be added in the future.
+func CollectRequestedKeys(expr syntax.Expr) ([]string, map[string]string) {
```
IMO this is closer to the domain of an optimization phase (like some kind of projection/predicate pushdown) that we do in the physical planner. 

I'm not sure we should do this here, especially since it requires deeper analysis of the entire LogQL expression.

ashwanthgoli replied:
+1 `pkg/engine/planner/physical/optimizer.go` might be a better place for this.
https://github.com/grafana/loki/blob/731ea38ac6b8d377c88858a4f45b5450aa0b8c0d/pkg/engine/planner/physical/optimizer.go#L140

For metric queries, we can apply the grouping columns columns from `RangeAggregation` as the requested keys.

rfratto commented on pkg/engine/planner/logical/builder_test.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+func TestBuilderParse(t *testing.T) {
+	t.Run("Builder adds Parse after MakeTable", func(t *testing.T) {
```
I don't think this test is actually validating any business logic. The assertion is "Builder adds Parse after MakeTable," but that's based on the test calling the builder in that order itself. 

We can remove this in favour of testing the planner API.

trevorwhitney replied:
agreed, it's just testing the test setup itself, will move planner tests higher through the the planning API

rfratto commented on pkg/engine/planner/logical/node_parse.go:
```diff
@@ -0,0 +1,62 @@
+
+	Table         Value // The table relation to parse from
+	Kind          ParserKind
+	RequestedKeys []string
+	NumericHints  map[string]NumericType // Fields that should be cast to numeric types
```
Both of these fields are closer to the domain of the physical planner and optimization passes where we can push down the requested keys / hints. 

I think we should take these out of the logical planner entirely.

rfratto commented on pkg/engine/planner/logical/node_parse_test.go:
```diff
@@ -0,0 +1,40 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseInstruction(t *testing.T) {
```
This test isn't testing any business logic at all? I think we can remove this.

rfratto commented on pkg/engine/planner/logical/planner_test.go:
```diff
@@ -257,3 +257,355 @@ func TestCanExecuteQuery(t *testing.T) {
+	})
+}
+
+func TestPlannerCreatesParseFromLogfmt(t *testing.T) {
```
Can you update all of these so they assert against the generated SSA like the other planner tests, please?

rfratto reviewed (commented):
I'll be doing a few reviews:

* Changes to logical planner (this review) 
* Changes to physical planner 
* Changes to query executor 
* Changes to the integration package 

I had partially reviewed the logical planner last week; my comments from then are still applicable to updating the logical planner.

rfratto reviewed (commented):
Here's my review for the changes to the query executor. 

I think we should take out unused code like the cast logic; there's a lot of other things that have actionable feedback and it's going to slow down your momentum if we scope unused code into the review. 

I'm also concerned about the logfmt tokenizer. It seems written by AI, and may have subtle bugs everywhere. Can we use an existing logfmt decoder instead of having AI write one, please?

rfratto reviewed (changes requested):
Thank you for your efforts here :)

This is introducing a very significant amount of code, and in my reviews, some of it feels out of scope for introducing logfmt parsing (casting), and other parts feel like we're adding new ways to do something we can already do (the column registry).  

Before we get this in main, I'd appreciate if we:

* Move logic for optimizing parsing into an optimization pass of the physical plan. 

* Reevaluate any instance where we introduce new constructs for things that should already be handled (i.e., column registry). 

* Prefer to test behaviour via the public API of packages, rather than testing internal details. This will give us more freedom to refactor without breaking tests. 

* Remove tests that don't touch business logic at all, since it's not adding any validations to the correctness of the code here. 

* Remove the logfmt tokenizer AI wrote and use an existing one. 

* Remove unused business logic (like the Cast node) or any nice-to-have additions (like integration testing)

trevorwhitney commented on pkg/engine/planner/physical/visitor_test.go:
```diff
@@ -17,6 +17,7 @@ type nodeCollectVisitor struct {
 	onVisitProjection        func(*Projection) error
 	onVisitRangeAggregation  func(*RangeAggregation) error
 	onVisitVectorAggregation func(*VectorAggregation) error
+	onVisitParse             func(*ParseNode) error
```
moving cast to a subsequent PR

trevorwhitney commented on pkg/engine/planner/physical/planner.go:
```diff
@@ -339,6 +367,62 @@ func (p *Planner) processVectorAggregation(lp *logical.VectorAggregation, ctx *C
+	}
+
+	// If there are numeric hints, insert a CastNode
+	if len(lp.NumericHints) > 0 {
```
moving numericHints to the subsequent cast PR

trevorwhitney commented:
@chaudum and @rfratto this should be ready for another look, I think I've addressed all the feedback. here's a summary of what's been done. thanks again for all your time on this!

* The big change was pushing the key collection out of the logical layer and into the physical by means of an optimization. I'll be honest I didn't fully understand the purpose of these optimizations, and figuring out keys to parse seemed like a logical step to me at first. However, thanks to your (and @ashwanthgoli's) thoughtful reviews I learned about optimizers and saw how that would be a better place, without the blending of concerns between metric and log queries I was introducing via collection in the logical layer.
* I removed all unwrap/cast code, this includes type hints, as it was too much scope for this first PR. I will follow up with that work later, since some of it is needed for Drilldown (cast, not unwrap).
* I refactored and consolidated tests to focus on testing through public APIs and removing any redundancy.
* I am using the existing logftmt decoder.
* I refactored the parse execution to fix the bug with different rows in the parsed values, as well as adopt Robert's suggestions regarding allocations as best I could. Since our decoder doesn't allow for early stopping, or requesting certain keys, we are still parsing the whole line, but we can add that behavior later in another PR.

The only feedback I wanted to push back on was the removal of the integration test. There are a lot of pieces to the query pipeline, and I honestly didn't know when I was done until I could actually issue a `logfmt` query and see it work.

> Especially since this seems beyond the scope of adding support for logfmt parsing.

I disagree. Yes, there are tests in there that cover other query paths, but I needed simpler paths to find and fix the bug in `sortmerge`, and there are also `logfmt` specific cases.

> While I appreciate you adding integration tests, this is quite a large file and I would like if we could take this out for now and rely more heavily on unit tests.

Would you be open to it if I refactored out the helper functions, or split it up into more test cases? It is big because we can't ingest via single binary, so there's a lot of scaffolding to setup the dataobjs, and that should be able to go away in the future, but in the meantime I find it the quickest way to see a query actually "work".

rfratto commented on integration/v2_query_engine_test.go:
Resolving since we've agreed there's value in the integration tests

trevorwhitney commented:
@rfratto nevermind. after rebasing the latest metastore changes into this branch the integration test is failing. I spent a good chunk of the day trying to figure out why and I can't, so I don' think it's worth the investment at this point. I may reintroduce it at a future point, but for the time being the benchmarks should be enough of a validation. I would be open to running those automatically for PRs that touch anything in the `engine` folder.

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -261,6 +269,144 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
 
 var _ rule = (*projectionPushdown)(nil)
 
+// parseKeysPushdown determines which keys need to be parsed based on downstream operations
+type parseKeysPushdown struct {
+	plan *Plan
+}
```
The keys to parse acts the same as a projection, doesn't it? Maybe we could wrap this into `applyProjectionPushdown` to avoid adding a new stage?

That would also help to get remove of all the `extract*` helper functions since you could just use the existing `projections` parameter in that function.

rfratto commented on pkg/engine/executor/executor_test.go:
```diff
@@ -97,3 +104,127 @@ func TestExecutor_Projection(t *testing.T) {
+				{"timestamp": time.Unix(2, 0).UTC(), "message": "level=info status=200", "app": "backend"},
+			},
+			expectedFields: 5, // 5 columns: timestamp, message, app, level, status
+			expectedParsedOutput: map[string][]string{
+				"level":  {"error", "info"},
+				"status": {"500", "200"},
+			},
```
I'd recommend changing these to use `arrowtest.Rows`:

```go
expectedParsedOutput: arrowtest.Rows{
  {"level": "error", "status": "500"}, 
  {"level: "info", "status": "200"}, 
}
```

then you can assert using:

```go
record, _ := pipeline.Value() 
defer record.Release()

actual, err := arrowtest.RecordRows(record)
require.NoError(t, err) 
require.Equal(t, tt.expectedParseOutput, actual) 
```

That would avoid the need for the getColumnIndices/verifyParseOutput helper functions, and generally row-order expctations are easier to read than column-order.

rfratto commented on pkg/engine/executor/parse.go:
```diff
@@ -0,0 +1,117 @@
+		}
+		for _, header := range headers {
+			// Add metadata to mark these as parsed columns
+			metadata := arrow.NewMetadata([]string{
+				types.MetadataKeyColumnType,
+				types.MetadataKeyColumnDataType,
+			}, []string{
+				types.ColumnTypeParsed.String(),
+				datatype.Loki.String.String(),
+			})
```
nit: use `datatype.ColumnMetadata(types.ColumnTypeParsed, datatype.Loki.String)` here instead

rfratto commented on pkg/engine/executor/parse.go:
```diff
@@ -0,0 +1,117 @@
+				types.ColumnTypeParsed.String(),
+				datatype.Loki.String.String(),
+			})
+			newFields = append(newFields, arrow.Field{
```
`Nullable: true` should be set here since there's going to be at least some null values if the set of keys isn't consistent across rows.

rfratto commented on pkg/engine/executor/parse_logfmt.go:
```diff
@@ -0,0 +1,189 @@
+		return result, err
+	}
+
+	// Filter to requested keys if specified
+	if len(requestedKeys) > 0 {
+		filteredResult := make(map[string]string)
+		for _, key := range requestedKeys {
+			if value, ok := result[key]; ok {
+				filteredResult[key] = value
+			}
+		}
+		return filteredResult, nil
+	}
```
You could save some memory here by creating the map prior to scanning, and then immediately skip over keys you don't want:

```go
var requestedLookup map[string]struct{} // built from requestedKeys 

for decoder.ScanKeyval() {
  key := string(decoder.Key())
  if requestedLookup != nil {
    _, want := requestedLookup[key] 
    if !want {
      continue 
    }
  }

  result[key] = string(decoder.Value()) 
}
```

rfratto commented on pkg/engine/executor/parse_logfmt.go:
```diff
@@ -0,0 +1,189 @@
+
+	// Parse all key-value pairs
+	for !decoder.EOL() && decoder.ScanKeyval() {
+		key := string(decoder.Key())
+		value := string(decoder.Value())
```
These will create a copy of the byte slice, you may want to use an `unsafeString` here to avoid the copy similar to what you do in creating the decoder.

rfratto commented on pkg/engine/executor/parse_logfmt.go:
```diff
@@ -0,0 +1,189 @@
+
+// parseAllKeys handles the dynamic path when no specific keys are requested
+// Discovers columns dynamically as lines are parsed
+func parseAllKeys(input *array.String, columnBuilders map[string]*array.StringBuilder, allocator memory.Allocator) ([]string, []error, bool) {
```
I wonder if we should combine parseAllKeys and parseRequestedKeys to avoid the double implementation? With a combined function, you'd still pass down the requested keys to the tokenizer, but the requested keys would be `nil` if all keys are needed.

I'm not sure the overhead of lazily creating builders and looking them up is high enough to make it worth having two implementations that we would have to keep in sync to fix bugs.

rfratto commented on pkg/engine/executor/parse_logfmt.go:
```diff
@@ -0,0 +1,189 @@
+	for i := 0; i < input.Len(); i++ {
+		line := input.Value(i)
+		parsed, err := TokenizeLogfmt(line, nil) // Get all keys
+		errorList = append(errorList, err)
+		if err != nil {
+			hasAnyError = true
+		}
```
nit: Instead of returning an error slice, maybe we want to bake in writing to the error column builder, since it's just another column? 

You'll know whether there were any errors by looking at `builder.Len() - builder.NullN()` to get the number of non-null rows.

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -261,6 +269,144 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
+}
+
+// apply implements rule.
+func (r *parseKeysPushdown) apply(node Node) bool {
```
I think the optimization pass here is going to break for logs queries, where they always need all keys. This may need similar logic for logs queries like what's done for the projection pushdown.

(But I think treating the parse key pushdown as a projection pushdown still makes sense, and then you could reuse the existing logic for handling logs queries there)

trevorwhitney commented on pkg/engine/planner/logical/planner.go:
```diff
@@ -135,9 +149,19 @@ func buildPlanForLogQuery(expr syntax.LogSelectorExpr, params logql.Params, isMe
 func buildPlanForSampleQuery(e syntax.SampleExpr, params logql.Params) (*Builder, error) {
-	var (
-		err error
+	if params.Step() > 0 {
+		return nil, fmt.Errorf("only instant metric queries are supported: %w", errUnimplemented)
+	}
```
yup, looks like I missed that one with the stream label detection bit, removing.

trevorwhitney commented on pkg/engine/planner/logical/planner.go:
```diff
@@ -100,6 +109,11 @@ func buildPlanForLogQuery(expr syntax.LogSelectorExpr, params logql.Params, isMe
 		},
 	)
 
+	// Add Parse instruction if logfmt was detected
+	if hasLogfmtParser {
+		builder = builder.Parse(ParserLogfmt)
+	}
```
I think it's reasonable to have the placement match. I will tackle this while fixing the bug regarding predicates in the Maketable, but with one exception, a line filter predicate will always come first (which is an optimization I think we do today in v1 anyway). Label filter predicates will be tracked pre and post filter, and only the pre-filter predicates (which will include line filters, so maybe that's a bad name?) will be included in the maketable predicates. the logfmt statement will come after all the pre-parse predicates, and the post-parse predicates will come after the parse statement.

trevorwhitney commented on pkg/engine/planner/logical/planner.go:
```diff
@@ -147,56 +171,64 @@ func buildPlanForSampleQuery(e syntax.SampleExpr, params logql.Params) (*Builder
 			for _, group := range e.Grouping.Groups {
-				groupBy = append(groupBy, *NewColumnRef(group, types.ColumnTypeAmbiguous))
+				// Determine column type based on whether it's a stream label
+				colType := types.ColumnTypeAmbiguous
+				if streamLabels[group] {
+					colType = types.ColumnTypeLabel
+				}
```
I think it's reasonable to want to do this in one place, and not make it the concern of the planner. I do think we should do it though. also, like you say, fixing the predicate order makes this unnecessary.

trevorwhitney commented on pkg/engine/planner/logical/planner.go:
```diff
@@ -147,56 +171,64 @@ func buildPlanForSampleQuery(e syntax.SampleExpr, params logql.Params) (*Builder
 
 	e.Walk(func(e syntax.Expr) bool {
```
I think you are correct, adding this in the first place was incorrect on my part, while it solved the problem, what was actually needed was moving up the order at which we walk a parse instruction, which was needed for correcting the predicates anyway.

rfratto commented on pkg/engine/planner/physical/planner.go:
```diff
@@ -363,8 +365,29 @@ func (p *Planner) processVectorAggregation(lp *logical.VectorAggregation, ctx *C
-// to the scan nodes.
+// Convert [logical.Parse] into one [ParseNode] node.
+func (p *Planner) processParse(lp *logical.Parse, ctx *Context) ([]Node, error) {
+	node := &ParseNode{
```
nit: it may be worth adding a comment explaining that RequestedKeys gets filled during an optimization pass

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -198,26 +199,54 @@ func (r *projectionPushdown) apply(node Node) bool {
-		return r.applyProjectionPushdown(node, projections, false)
+		ambiguousProjections := collectAmbiguousColumns(node.PartitionBy)
+		return r.applyProjectionPushdown(node, projections, ambiguousProjections, false)
+	case *VectorAggregation:
```
cc @ashwanthgoli Is there a reason we didn't have pushdown for VectorAggregation before?

rfratto replied:
Ah, okay. Do you have a suggestion for what we should do in the projection pushdown for the parse?

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -198,26 +199,54 @@ func (r *projectionPushdown) apply(node Node) bool {
+		ambiguousProjections := collectAmbiguousColumns(node.PartitionBy)
+		return r.applyProjectionPushdown(node, projections, ambiguousProjections, false)
+	case *VectorAggregation:
+		if len(node.GroupBy) == 0 || node.Operation != types.VectorAggregationTypeSum {
```
I guess this is copied from the RangeAggregation, but I'm not sure why we only allow it to pushdown predicates for Count/Sum and none of the others? 

cc @ashwanthgoli

trevorwhitney replied:
we have a check earlier in the pipeline that returns an unimplemented for anything other than sum by count_over_time, so my assumption was this was a byproduct of that, and meant to be temporary until we implement it?

that being said, in order to support drilldown we'll need to support avg_over_time, which is just a sum/count, so since we already have count I think we should implement sum next.

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -198,26 +199,54 @@ func (r *projectionPushdown) apply(node Node) bool {
+func (r *projectionPushdown) applyProjectionPushdown(
+	node Node,
+	projections []ColumnExpression,
+	ambiguousProjections []ColumnExpression,
```
nit: rather than passing ambiguousProjections as an arugment (which is always a subset of projections but not used for DataObjScan), I think the `ParseNode` case should build the subset itself to keep the number of parameters minimal.

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -248,19 +277,86 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
+				continue
+			}
+
+			// Only collect ambiguous columns to push to parse nodes
+			if colExpr.Ref.Type == types.ColumnTypeAmbiguous && !existingKeys[colExpr.Ref.Column] {
+				existingKeys[colExpr.Ref.Column] = true
+				changed = true
+			}
```
nit: this can be simplified a bit to just setting `existingKeys[colExpr.Ref.Column] = true`, since the projections are always ambiguous. Then, after the loop, you can initialize changed to `changed = len(existingKeys) > len(node.RequestedKeys)`.

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -248,19 +277,86 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
+		}
+
+		// Found a ParseNode - update its keys
+		existingKeys := make(map[string]bool)
```
nit: existingKeys is maybe misleading here since it's a combination of the old existing keys and the new projected keys.

Maybe rename to just `requestedKeys`?

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -248,19 +277,86 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
+		}
+
+		if changed {
+			// Convert back to sorted slice
+			newKeys := make([]string, 0, len(existingKeys))
+			for k := range existingKeys {
+				newKeys = append(newKeys, k)
+			}
+			sort.Strings(newKeys)
```
nit: this can be modernized with 

```go
newKeys := slices.Collect(maps.Keys(existingKeys)) 
sort.Strings(newKeys) 
```

rfratto commented on pkg/engine/planner/physical/optimizer.go:
```diff
@@ -248,19 +277,86 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
 		}
 		return changed
+	case *ParseNode:
+		// Only apply the pushdown for Metric queries. Log queries should request all keys
+		if !r.isMetricQuery() || len(ambiguousProjections) == 0 {
+			return false
+		}
```
> Log queries should request all keys

This may not be true forever, so I don't think we want to enforce that logic here. The DataObjScan case uses `len(node.Projections) == 0 && applyIfNotEmpty` to represent this logic; can we use the same here? 

(Then we could remove isMetricQuery)

rfratto commented on pkg/engine/planner/physical/optimizer_test.go:
There's two big cases I think we want to test here: 

1. Ambiguous projections pushed down to ParseNode should not also propagate down to the DataObjScan (e.g., the projections get "consumed" by ParseNode) 

2. Predicates do not push down past a ParseNode into a DataObjScan

rfratto commented on pkg/engine/executor/executor_test.go:
Much easier to read!

rfratto commented on pkg/engine/executor/executor_test.go:
```diff
@@ -97,3 +102,181 @@ func TestExecutor_Projection(t *testing.T) {
+		t.Run(tt.name, func(t *testing.T) {
+			// Create input data with message column containing logfmt
+			input := NewArrowtestPipeline(
+				memory.DefaultAllocator,
```
nit: You may with to create a checked allocator here to assert that we do releases correctly:

```go
alloc := memory.NewCheckedAllocator(memory.DefaultAllocator) 
defer alloc.AssertSize(t, 0) // Assert empty on test exit 

input := NewArrowtestPipeline(alloc, tt.schema, tt.input) 

// ... 

pipeline := NewParsePipeline(parseNode, input, alloc) 
```

rfratto commented on pkg/engine/executor/parse_logfmt.go:
```diff
@@ -0,0 +1,153 @@
+// TokenizeLogfmt parses logfmt input using the standard decoder
+// Returns a map of key-value pairs with last-wins semantics for duplicates
+// If requestedKeys is provided, the result will be filtered to only include those keys
+func TokenizeLogfmt(input string, requestedKeys []string) (map[string]string, error) {
```
nit: this probably doesn't need to be exported as part of the public API

rfratto commented on pkg/engine/executor/parse_logfmt.go:
```diff
@@ -0,0 +1,153 @@
+				columnOrder = append(columnOrder, "__error__", "__error_details__")
+				hasErrorColumns = true
+
+				// Backfill NULLs for previous rows
+				for j := 0; j < i; j++ {
+					errorBuilder.AppendNull()
+					errorDetailsBuilder.AppendNull()
+				}
```
small nit: you could use `AppendNulls(n)` here to avoid the loop. (ditto in other places where you append nulls in a loop)

trevorwhitney replied:
if you feel strongly about it for readability, but each call to `AppendNulls` will create it's own loop, so when multiple columns need nulls appended doing it this way is slightly more performant, though it's probably marginal.

rfratto replied:
> if you feel strongly about it 

nah, just a small nit :) I'm fine with it as-is.

rfratto commented:
I just fired off a correctness test run for final validation: https://github.com/grafana/loki/actions/runs/17766971412

rfratto commented:
Second attempt: https://github.com/grafana/loki/actions/runs/17767292602
